icu4c/source/i18n/formatted_string_builder.cpp

```diff
@@ -275,14 +297,19 @@ int32_t FormattedStringBuilder::prepareForInsertHelper(int32_t index, int32_t co
     int32_t oldZero = fZero;
     char16_t *oldChars = getCharPtr();
     Field *oldFields = getFieldPtr();
+    int32_t newLength;
+    if (uprv_add32_overflow(fLength, count, &newLength)) {
+        status = U_INPUT_TOO_LONG_ERROR;
+        return -1;
+    }
     if (fLength + count > oldCapacity) {
-        if ((fLength + count) > INT32_MAX / 2) {
-            // If we continue, then newCapacity will overflow int32_t in the next line.
+        int32_t newCapacity;
+        //  newCapacity = (fLength + count) * 2 w/ overflow checking.
+        if (uprv_add32_overflow(newLength, newLength, &newCapacity)) {
             status = U_INPUT_TOO_LONG_ERROR;
             return -1;
         }
-        int32_t newCapacity = (fLength + count) * 2;
-        int32_t newZero = newCapacity / 2 - (fLength + count) / 2;
+        int32_t newZero = newLength / 2;
```

Comment:
```suggestion
        int32_t newZero = (newCapacity - newLength) / 2;
```
... to center the contents in the buffer, as before.